web/interceptor/tokens_test.go

```diff
@@ -0,0 +1,159 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lis := bufconn.Listen(BufSize)
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return lis.Dial()
+	}
+	opt := grpc.ChainUnaryInterceptor(
+		interceptor.UnaryUserVerifier(logger, tm),
+		interceptor.TokenRefresher(logger, tm),
+	)
+	s := grpc.NewServer(opt) // skipcq: GO-S0902
+	qf.RegisterQuickFeedServiceServer(s, ags)
+
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			t.Errorf("Server exited with error: %v", err)
+			return
+		}
+	}()
+
+	admin := qtest.CreateFakeUser(t, db, 1)
+	user := qtest.CreateFakeUser(t, db, 56)
+
+	ctx := context.Background()
+	conn, err := grpc.DialContext(ctx, "bufnet", grpc.WithContextDialer(bufDialer), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial bufnet: %v", err)
+	}
+	defer conn.Close()
+
+	client := qf.NewQuickFeedServiceClient(conn)
+
+	adminCookie, err := tm.NewAuthCookie(admin.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	userCookie, err := tm.NewAuthCookie(user.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	adminCtx := qtest.WithAuthCookie(ctx, interceptor.AuthTokenString(adminCookie.Value))
+	userCtx := qtest.WithAuthCookie(ctx, interceptor.AuthTokenString(userCookie.Value))
+	adminClaims := &auth.Claims{
+		UserID: admin.ID,
+		Admin:  true,
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(1 * time.Minute).Unix(),
+		},
+	}
+	userClaims := &auth.Claims{
+		UserID: user.ID,
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(1 * time.Minute).Unix(),
+		},
+	}
+	if tm.UpdateRequired(adminClaims) || tm.UpdateRequired(userClaims) {
+		t.Error("No users should be in the token update list at the start")
+	}
+	if _, err := client.GetUsers(adminCtx, &qf.Void{}); err != nil {
+		t.Fatal(err)
+	}
+	if tm.UpdateRequired(userClaims) {
+		t.Error("No users should be in the token update list")
+	}
+	if _, err := client.UpdateUser(adminCtx, user); err != nil {
+		t.Fatal(err)
+	}
+	if !tm.UpdateRequired(userClaims) {
+		t.Error("User must be in the token update list")
+	}
+	if _, err := client.GetUser(userCtx, &qf.Void{}); err != nil {
+		t.Fatal(err)
+	}
+	if tm.UpdateRequired(userClaims) {
+		t.Error("User should not be in the token update list after the token has been updated")
+	}
+	course := &qf.Course{
+		ID:               1,
+		OrganizationID:   1,
+		OrganizationPath: "testorg",
+		Provider:         "fake",
+	}
+	group := &qf.Group{
+		ID:       1,
+		Name:     "test",
+		CourseID: 1,
+		Users: []*qf.User{
+			user,
+		},
+	}
+	if _, err := client.CreateCourse(adminCtx, course); err != nil {
+		t.Fatal(err)
+	}
+	if !tm.UpdateRequired(adminClaims) {
+		t.Error("Admin must be in the token update list after creating a new course")
+	}
+	qtest.EnrollStudent(t, db, user, course)
+	if _, err := client.CreateGroup(adminCtx, group); err != nil {
+		t.Fatal(err)
+	}
+	if tm.UpdateRequired(userClaims) {
+		t.Error("User should not be in the token update list")
```

Comment:
I guess it would be helpful to add context similar to 126 above, e.g., `... after enrolling in course`. Should the check be done right after EnrollStudent on line 128?

Reply:
`EnrollStudent` does not affect the token update list as it is performed directly on the database, but `CreateGroup` shouldn't and must be checked.